std-bits/table/src/main/java/org/enso/table/parsing/NumberParser.java

```diff
@@ -0,0 +1,279 @@
+    private final boolean allowScientific;
+    private final boolean trimValues;
+    private final String separators;
+
+    /**
+     * Creates a new integer instance of this parser.
+     *
+     * @param allowCurrency whether to allow currency symbols
+     * @param allowLeadingZeros whether to allow leading zeros
+     * @param trimValues whether to trim the input values
+     * @param thousandSeparator the thousand separator to use
+     */
+    public NumberParser(boolean allowCurrency, boolean allowLeadingZeros, boolean trimValues, String thousandSeparator) {
+        this.allowCurrency = allowCurrency;
+        this.allowLeadingZeros = allowLeadingZeros;
+        this.trimValues = trimValues;
+
+        this.allowDecimal = false;
+        this.allowScientific = false;
+
+        this.separators = thousandSeparator == null ? null : (thousandSeparator + '_');
+    }
+
+    /**
+     * Creates a new decimal instance of this parser.
+     *
+     * @param allowCurrency whether to allow currency symbols
+     * @param allowLeadingZeros whether to allow leading zeros
+     * @param trimValues whether to trim the input values
+     * @param allowScientific whether to allow scientific notation
+     */
+    public NumberParser(boolean allowCurrency, boolean allowLeadingZeros, boolean trimValues, boolean allowScientific) {
+        this.allowCurrency = allowCurrency;
+        this.allowLeadingZeros = allowLeadingZeros;
+        this.trimValues = trimValues;
+
+        this.allowDecimal = true;
+        this.allowScientific = allowScientific;
+
+        this.separators = null;
+    }
+
+    /**
+     * Creates a new decimal instance of this parser.
+     *
+     * @param allowCurrency whether to allow currency symbols
+     * @param allowLeadingZeros whether to allow leading zeros
+     * @param trimValues whether to trim the input values
+     * @param allowScientific whether to allow scientific notation
+     * @param thousandSeparator the thousand separator to use
+     * @param decimalSeparator the decimal separator to use
+     */
+    public NumberParser(boolean allowCurrency, boolean allowLeadingZeros, boolean trimValues, boolean allowScientific, String thousandSeparator, String decimalSeparator) {
+        if (decimalSeparator.length() != 1) {
+            throw new IllegalArgumentException("Decimal separator must be a single character.");
+        }
+
+        this.allowCurrency = allowCurrency;
+        this.allowLeadingZeros = allowLeadingZeros;
+        this.trimValues = trimValues;
+
+        this.allowDecimal = true;
+        this.allowScientific = allowScientific;
+
+        this.separators = thousandSeparator + decimalSeparator;
+    }
+
+    private Pattern patternForIndex(int index) {
+        return separators == null
+            ? getPattern(allowDecimal, allowCurrency, allowScientific, trimValues, index)
+            : getPattern(allowDecimal, allowCurrency, allowScientific, trimValues, index, separators);
+    }
```

Comment:
I think I don't understand the concept of pattern `index`. What does it mean??

I assume we just have a sequence of different numeric patterns that we try to fit, is that right? I'd appreciate some explanation.